File: apps/android/app/src/main/java/com/kelpie/browser/ui/AccountButton.kt
package com.kelpie.browser.ui

import android.app.Activity
import android.graphics.BitmapFactory
import androidx.compose.foundation.Image
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.PersonOutline
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.asImageBitmap
import androidx.compose.ui.unit.dp
import com.kelpie.browser.account.UOAAccount
import com.kelpie.browser.browser.BookmarkStore

@Composable
internal fun AccountButton(activity: Activity) {
    val account by UOAAccount.state.collectAsState()
    val syncError by BookmarkStore.syncError.collectAsState()
    var menu by remember { mutableStateOf(false) }
    Box {
        IconButton(onClick = {
            if (account.profile == null && !account.signingIn) UOAAccount.signIn() else menu = true
        }, modifier = Modifier.size(44.dp)) {
            val avatar = remember(account.avatar) { account.avatar?.let { decodeAvatar(it) } }
            val label = if (account.profile == null) "Login/register" else "Account"
            if (avatar != null) {
                Image(avatar.asImageBitmap(), label, Modifier.size(32.dp).clip(CircleShape))
            } else {
                Icon(Icons.Default.PersonOutline, label)
            }
        }
        DropdownMenu(menu, onDismissRequest = { menu = false }) {
            if (account.signingIn) {
                DropdownMenuItem(text = { Text("Cancel login") }, onClick = {
                    menu = false
                    UOAAccount.signOut()
                })
            } else {
                DropdownMenuItem(text = { Text(account.profile?.name ?: account.profile?.email.orEmpty()) }, enabled = false, onClick = {})
                if (syncError != null) DropdownMenuItem(text = { Text(syncError.orEmpty()) }, enabled = false, onClick = {})
                DropdownMenuItem(text = { Text("Refresh favourites") }, onClick = {
                    menu = false
                    BookmarkStore.refreshAccountBookmarks()
                })
                DropdownMenuItem(text = { Text("Sign out") }, onClick = {
                    menu = false
                    UOAAccount.signOut()
                })
            }
        }
    }
    account.error?.let { message ->
        AlertDialog(
            onDismissRequest = UOAAccount::dismissError,
            text = { Text(message) },
            confirmButton = { TextButton(onClick = UOAAccount::dismissError) { Text("OK") } },
        )
    }
}

private fun decodeAvatar(data: ByteArray): android.graphics.Bitmap? {
    val options = BitmapFactory.Options().apply { inJustDecodeBounds = true }
    BitmapFactory.decodeByteArray(data, 0, data.size, options)
    options.inSampleSize = maxOf(1, maxOf(options.outWidth, options.outHeight) / 128)
    options.inJustDecodeBounds = false
    return BitmapFactory.decodeByteArray(data, 0, data.size, options)
}

File: apps/ios/Kelpie/Views/AccountButton.swift
import SwiftUI

struct AccountButton: View {
    @ObservedObject private var account = UOAAccount.shared
    @ObservedObject private var bookmarks = BookmarkStore.shared
    @State private var showMenu = false

    var body: some View {
        Button {
            if account.profile == nil && !account.signingIn { account.signIn() }
            else { showMenu = true }
        } label: {
            Group {
                if let avatar = account.avatar {
                    Image(uiImage: avatar).resizable().scaledToFill()
                        .frame(width: 32, height: 32).clipShape(Circle())
                } else {
                    Image(systemName: "person.crop.circle").font(.system(size: 22))
                }
            }
            .frame(width: 44, height: 44)
            .modifier(BrowserGlass())
        }
        .accessibilityLabel(account.profile == nil ? "Login/register" : "Account")
        .accessibilityIdentifier("browser.account")
        .confirmationDialog("Account", isPresented: $showMenu, titleVisibility: .visible) {
            if account.signingIn {
                Button("Cancel login", action: account.cancelSignIn)
            } else {
                Button("Refresh favourites", action: bookmarks.refreshAccountBookmarks)
                Button("Sign out", role: .destructive, action: account.signOut)
            }
        } message: {
            Text(bookmarks.syncError ?? account.profile?.name ?? account.profile?.email ?? "")
        }
        .alert("Login", isPresented: Binding(get: { account.error != nil }, set: { if !$0 { account.error = nil } })) {
            Button("OK") { account.error = nil }
        } message: {
            Text(account.error ?? "")
        }
    }
}
